cylc/flow/data_store_mgr.py

```diff
@@ -744,146 +741,343 @@ def increment_graph_window(
+                                self.generate_edge(
+                                    parent_tokens,
+                                    node_tokens,
+                                    active_id
+                                )
+                                np_ids.add(parent_tokens.id)
+
+                    # Register new walk
+                    if node_id not in all_walks:
+                        all_walks[node_id] = {
+                            'locations': {},
+                            'done_ids': set(),
+                            'done_locs': set(),
+                            'orphans': set(),
+                            'walk_ids': {node_id} | nc_ids | np_ids,
+                            'depths': {
+                                depth: set()
+                                for depth in range(1, self.n_edge_distance + 1)
+                            }
+                        }
+                    if nc_ids:
+                        all_walks[node_id]['locations'][c_tag] = nc_ids
+                        all_walks[node_id]['depths'][1].update(nc_ids)
+                        c_ids.update(nc_ids)
+                    if np_ids:
+                        all_walks[node_id]['locations'][p_tag] = np_ids
+                        all_walks[node_id]['depths'][1].update(np_ids)
+                        p_ids.update(np_ids)
+
+                # Create location association
+                c_ids.difference_update(active_walk['walk_ids'])
+                if c_ids:
+                    active_locs.setdefault(c_loc, set()).update(c_ids)
+                p_ids.difference_update(active_walk['walk_ids'])
+                if p_ids:
+                    active_locs.setdefault(p_loc, set()).update(p_ids)
+                active_walk['walk_ids'].update(c_ids, p_ids)
+                active_walk['depths'][n_depth].update(c_ids, p_ids)
+
+        self.n_window_completed_walks.add(active_id)
+        self.n_window_nodes[active_id].update(active_walk['walk_ids'])
 
-        # If this is the active task (edge_distance has been incremented),
-        # then add the most distant child as a trigger to prune it.
-        if is_active:
-            levels = self.n_window_boundary_nodes[active_id].keys()
+        # This part is vital to constructing a set of boundary nodes
+        # associated with the n=0 window of current active node.
+        # Only trigger pruning for furthest set of boundary nodes
+        boundary_nodes: Set[str] = set()
+        max_level: int = 0
+        with suppress(ValueError):
+            max_level = max(
+                len(loc)
+                for loc in active_locs
+                if p_tag not in loc
+            )
+            # add the most distant child as a trigger to prune it.
+            boundary_nodes.update(*(
+                active_locs[loc]
+                for loc in active_locs
+                if p_tag not in loc and len(loc) >= max_level
+            ))
```

Comment:
Where is the `ValueError` expected to occur here?

Reply:
This is from the `max`:
```
>>> max(blah for blah in set())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: max() arg is an empty sequence
```
which can happen with isolates and end-of-branch nodes.

Reply:
I would suggest splitting it to avoid potentially catching a ValueError in the next bit
```suggestion
        try:
            max_level = max(
                len(loc)
                for loc in active_locs
                if p_tag not in loc
            )
        except ValueError:
            pass
        else:
            # add the most distant child as a trigger to prune it.
            boundary_nodes.update(*(
                active_locs[loc]
                for loc in active_locs
                if p_tag not in loc and len(loc) >= max_level
            ))
```